### PR TITLE
Add support for /hierarchy API and fall back to /spaces API if needed

### PR DESCRIPTION
**Pull request by t3chguy**

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -8008,6 +8009,60 @@ export class MatrixClient extends EventEmitter {
...
 
+    /**
+     * Fetches or paginates a room hierarchy as defined by MSC2946.
+     * Falls back gracefully to sourcing its data from `getSpaceSummary` if this API is not yet supported by the server.
+     * @param {string} roomId The ID of the space-room to use as the root of the summary.
+     * @param {number?} limit The maximum number of rooms to return per page.
+     * @param {number?} maxDepth The maximum depth in the tree from the root room to return.
+     * @param {boolean?} suggestedOnly Whether to only return rooms with suggested=true.
+     * @param {string?} fromToken The opaque token to paginate a previous request.
+     * @returns {Promise} the response, with next_token & rooms fields.
```
Comment (clokep) [inline src/client.ts:8020]: We updated this to `next_batch` in matrix-org/synapse#10579!

File: src/@types/spaces.ts [hunk 2]
```diff
@@ -15,26 +15,46 @@ limitations under the License.
...
 // Types relating to Rooms of type `m.space` and related APIs
 
 /* eslint-disable camelcase */
+/** @deprecated */
 export interface ISpaceSummaryRoom extends IPublicRoomsChunkRoom {
     num_refs: number;
     room_type: string;
 }
 
+/** @deprecated */
```
Comment (turt2live) [inline src/@types/spaces.ts:30]:
```suggestion
/** @deprecated Use hierarchy instead where possible. */
```

File: src/@types/spaces.ts [hunk 3]
```diff
@@ -15,26 +15,46 @@ limitations under the License.
...
 export interface ISpaceSummaryEvent {
```
Comment (turt2live) [inline src/@types/spaces.ts:31]: I'm surprised we don't have a generic event interface yet

Comment (t3chguy) [reply src/@types/spaces.ts:31]: We do, this predates it, keeping the identifier until deprecation expires and this gets canned

**Pull request merged**